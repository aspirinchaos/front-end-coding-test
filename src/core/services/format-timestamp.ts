const formatTimestamp = (timestamp: number): string => {
  return new Date(timestamp * 1000).toLocaleDateString('en-US', {
    year: 'numeric',
    month: 'long',
    day: 'numeric',
  });
};

export default formatTimestamp;
