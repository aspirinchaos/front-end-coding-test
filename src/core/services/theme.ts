import { DefaultTheme, IBreakpoint } from 'styled-components';

import { min, max } from './styled-responsive';

const BREAKPOINTS: IBreakpoint = {
  xs: 576,
  sm: 600,
  md: 992,
  lg: 1200,
  xl: 1700,
};

const theme: DefaultTheme = {
  breakpoints: BREAKPOINTS,
  colors: {
    main: '#e85a4f',
    secondary: '#8e8d8a',
    body: '#eae7dc',
  },
  font: 'Poppins,sans-serif',
  min,
  max,
};

export default theme;
