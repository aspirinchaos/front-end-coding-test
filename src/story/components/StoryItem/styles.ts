import styled, { css } from 'styled-components';

export const StoryContainer = styled.div`
  display: flex;
  width: 50%;
  ${(p) => p.theme.max('md')`
    width: 100%;
  `};
  ${(p) => p.theme.min('xl')`
    width: 33%;
  `};
  ${(p) => p.theme.max('sm')`
    flex-wrap: wrap;
  `};
`;

export const Image = styled.img`
  max-width: 100%;
  height: auto;
`;

const side = css`
  width: 50%;
  ${(p) => p.theme.max('sm')`
    width: 100%;
  `};
`;

export const ImageSide = styled.div`
  padding: 5px;
  ${side};
`;

export const TextSide = styled.div`
  padding: 15px;
  ${side};
`;

export const Title = styled.a`
  color: ${(p) => p.theme.colors.main};
  font-size: 1.125rem;
  text-decoration: none;
  font-family: ${(p) => p.theme.font};
  font-weight: 700;
  display: block;
  margin-bottom: 10px;

  &:hover {
    text-decoration: underline;
  }
`;

export const Info = styled.div`
  color: ${(p) => p.theme.colors.secondary};
  font-family: ${(p) => p.theme.font};
  font-size: 0.875rem;
  font-weight: 300;
  margin-bottom: 5px;
`;
