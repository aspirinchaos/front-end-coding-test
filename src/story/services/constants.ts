export const TOP_STORIES_URL =
  'https://hacker-news.firebaseio.com/v0/topstories.json';

export const STORY_ITEM_URL = (id: number): string =>
  `https://hacker-news.firebaseio.com/v0/item/${id}.json`;

export const USER_URL = (id: string): string =>
  ` https://hacker-news.firebaseio.com/v0/user/${id}.json`;
