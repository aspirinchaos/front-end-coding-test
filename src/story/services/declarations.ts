export interface StoryItem {
  by: string;
  score: number;
  title: string;
  url: string;
  time: number;
}

export interface User {
  id: string;
  karma: number;
}

export interface Story {
  id: number;
  title: string;
  url: string;
  timestamp: number;
  score: number;
  authorId: string;
  authorKarma: number;
}
