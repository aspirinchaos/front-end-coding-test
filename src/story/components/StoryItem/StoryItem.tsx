import React from 'react';

import IMAGE from 'core/assets/image.png';
import { Story } from 'story/services/declarations';

import { Flex, CalendarIcon, KarmaIcon, StarIcon, UserIcon } from 'core/atoms';
import formatTimestamp from 'core/services/format-timestamp';
import {
  StoryContainer,
  Image,
  ImageSide,
  TextSide,
  Title,
  Info,
} from './styles';

interface StoryItemProps {
  story: Story;
}

const StoryItem: React.FC<StoryItemProps> = ({ story }) => (
  <StoryContainer>
    <ImageSide>
      <Image src={IMAGE} />
    </ImageSide>
    <TextSide>
      <Title href={story.url} target={'_blank'}>
        {story.title}
      </Title>
      <Flex justifyContent={'space-between'}>
        <Info>
          <CalendarIcon /> {formatTimestamp(story.timestamp)}
        </Info>
        <Info>
          <StarIcon /> {story.score}
        </Info>
      </Flex>
      <Flex justifyContent={'space-between'}>
        <Info>
          <UserIcon /> {story.authorId}
        </Info>
        <Info>
          <KarmaIcon /> {story.authorKarma}
        </Info>
      </Flex>
    </TextSide>
  </StoryContainer>
);

export default StoryItem;
