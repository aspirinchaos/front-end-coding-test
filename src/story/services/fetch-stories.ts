import { toast } from 'react-toastify';

import { TOP_STORIES_URL, STORY_ITEM_URL, USER_URL } from './constants';
import { Story, StoryItem, User } from './declarations';

const fetch10Stories = async (): Promise<number[]> => {
  const response = await fetch(TOP_STORIES_URL);
  const data = await response.json();
  return data.sort(() => 0.5 - Math.random()).slice(0, 10);
};

const fetchStoryItem = async (id: number): Promise<StoryItem> => {
  const response = await fetch(STORY_ITEM_URL(id));
  return response.json();
};

const fetchUser = async (id: string): Promise<User> => {
  const response = await fetch(USER_URL(id));
  return response.json();
};

const fetchStories = async (): Promise<Story[]> => {
  try {
    const storiesIds = await fetch10Stories();
    const stories = await Promise.all(
      storiesIds.map(async (id) => {
        const story = await fetchStoryItem(id);
        const author = await fetchUser(story.by);

        return {
          id,
          title: story.title,
          url: story.url,
          timestamp: story.time,
          score: story.score,
          authorId: story.by,
          authorKarma: author.karma,
        };
      }),
    );

    return stories.sort((a, b) => a.score - b.score);
  } catch (error) {
    toast.error('Something went wrong with API');
    return [];
  }
};

export default fetchStories;
