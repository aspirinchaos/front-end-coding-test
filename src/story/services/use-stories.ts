import { useEffect, useState } from 'react';

import { Story } from 'story/services/declarations';
import fetchStories from 'story/services/fetch-stories';

const useStories = (): Story[] => {
  const [stories, setStories] = useState<Story[]>([]);

  useEffect(() => {
    const fetchData = async () => {
      const data = await fetchStories();
      setStories(data);
    };

    fetchData();
  }, []);

  return stories;
};

export default useStories;
