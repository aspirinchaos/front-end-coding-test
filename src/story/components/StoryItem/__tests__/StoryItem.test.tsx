import React from 'react';
import { render } from '@testing-library/react';
import 'jest-styled-components';
import { ThemeProvider } from 'styled-components';

import theme from 'core/services/theme';
import { Story } from 'story/services/declarations';

import StoryItem from '..';

const mockStory: Story = {
  authorId: 'authorId',
  authorKarma: 1000,
  id: 77777777,
  score: 100,
  timestamp: 1629888872,
  title: 'Test story item',
  url: 'https://test.story/article',
};

describe('StoryItem component', () => {
  it('renders correctly', () => {
    const { container } = render(
      <ThemeProvider theme={theme}>
        <StoryItem story={mockStory} />
      </ThemeProvider>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
