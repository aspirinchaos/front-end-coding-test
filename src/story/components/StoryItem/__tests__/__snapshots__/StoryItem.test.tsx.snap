// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StoryItem component renders correctly 1`] = `
.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 50%;
}

.c2 {
  max-width: 100%;
  height: auto;
}

.c1 {
  padding: 5px;
  width: 50%;
}

.c3 {
  padding: 15px;
  width: 50%;
}

.c4 {
  color: #e85a4f;
  font-size: 1.125rem;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-family: Poppins,sans-serif;
  font-weight: 700;
  display: block;
  margin-bottom: 10px;
}

.c4:hover {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c6 {
  color: #8e8d8a;
  font-family: Poppins,sans-serif;
  font-size: 0.875rem;
  font-weight: 300;
  margin-bottom: 5px;
}

@media (max-width:992px) {
  .c0 {
    width: 100%;
  }
}

@media (min-width:1700px) {
  .c0 {
    width: 33%;
  }
}

@media (max-width:600px) {
  .c0 {
    -webkit-flex-wrap: wrap;
    -ms-flex-wrap: wrap;
    flex-wrap: wrap;
  }
}

@media (max-width:600px) {
  .c1 {
    width: 100%;
  }
}

@media (max-width:600px) {
  .c3 {
    width: 100%;
  }
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <img
      class="c2"
      src="image.png"
    />
  </div>
  <div
    class="c3"
  >
    <a
      class="c4"
      href="https://test.story/article"
      target="_blank"
    >
      Test story item
    </a>
    <div
      class="c5"
    >
      <div
        class="c6"
      >
        <svg
          fill="currentColor"
          height="1em"
          version="1.1"
          viewBox="0 0 448 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M148 288h-40c-6.6 0-12-5.4-12-12v-40c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v40c0 6.6-5.4 12-12 12zm108-12v-40c0-6.6-5.4-12-12-12h-40c-6.6 0-12 5.4-12 12v40c0 6.6 5.4 12 12 12h40c6.6 0 12-5.4 12-12zm96 0v-40c0-6.6-5.4-12-12-12h-40c-6.6 0-12 5.4-12 12v40c0 6.6 5.4 12 12 12h40c6.6 0 12-5.4 12-12zm-96 96v-40c0-6.6-5.4-12-12-12h-40c-6.6 0-12 5.4-12 12v40c0 6.6 5.4 12 12 12h40c6.6 0 12-5.4 12-12zm-96 0v-40c0-6.6-5.4-12-12-12h-40c-6.6 0-12 5.4-12 12v40c0 6.6 5.4 12 12 12h40c6.6 0 12-5.4 12-12zm192 0v-40c0-6.6-5.4-12-12-12h-40c-6.6 0-12 5.4-12 12v40c0 6.6 5.4 12 12 12h40c6.6 0 12-5.4 12-12zm96-260v352c0 26.5-21.5 48-48 48H48c-26.5 0-48-21.5-48-48V112c0-26.5 21.5-48 48-48h48V12c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v52h128V12c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v52h48c26.5 0 48 21.5 48 48zm-48 346V160H48v298c0 3.3 2.7 6 6 6h340c3.3 0 6-2.7 6-6z"
          />
        </svg>
         
        August 25, 2021
      </div>
      <div
        class="c6"
      >
        <svg
          fill="currentColor"
          height="1em"
          version="1.1"
          viewBox="0 0 576 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M259.3 17.8L194 150.2 47.9 171.5c-26.2 3.8-36.7 36.1-17.7 54.6l105.7 103-25 145.5c-4.5 26.3 23.2 46 46.4 33.7L288 439.6l130.7 68.7c23.2 12.2 50.9-7.4 46.4-33.7l-25-145.5 105.7-103c19-18.5 8.5-50.8-17.7-54.6L382 150.2 316.7 17.8c-11.7-23.6-45.6-23.9-57.4 0z"
          />
        </svg>
         
        100
      </div>
    </div>
    <div
      class="c5"
    >
      <div
        class="c6"
      >
        <svg
          fill="currentColor"
          height="1em"
          version="1.1"
          viewBox="0 0 640 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M224 256c70.7 0 128-57.3 128-128S294.7 0 224 0 96 57.3 96 128s57.3 128 128 128zm89.6 32h-16.7c-22.2 10.2-46.9 16-72.9 16s-50.6-5.8-72.9-16h-16.7C60.2 288 0 348.2 0 422.4V464c0 26.5 21.5 48 48 48h274.9c-2.4-6.8-3.4-14-2.6-21.3l6.8-60.9 1.2-11.1 7.9-7.9 77.3-77.3c-24.5-27.7-60-45.5-99.9-45.5zm45.3 145.3l-6.8 61c-1.1 10.2 7.5 18.8 17.6 17.6l60.9-6.8 137.9-137.9-71.7-71.7-137.9 137.8zM633 268.9L595.1 231c-9.3-9.3-24.5-9.3-33.8 0l-37.8 37.8-4.1 4.1 71.8 71.7 41.8-41.8c9.3-9.4 9.3-24.5 0-33.9z"
          />
        </svg>
         
        authorId
      </div>
      <div
        class="c6"
      >
        <svg
          fill="currentColor"
          height="1em"
          version="1.1"
          viewBox="0 0 384 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M97.12 362.63c-8.69-8.69-4.16-6.24-25.12-11.85-9.51-2.55-17.87-7.45-25.43-13.32L1.2 448.7c-4.39 10.77 3.81 22.47 15.43 22.03l52.69-2.01L105.56 507c8 8.44 22.04 5.81 26.43-4.96l52.05-127.62c-10.84 6.04-22.87 9.58-35.31 9.58-19.5 0-37.82-7.59-51.61-21.37zM382.8 448.7l-45.37-111.24c-7.56 5.88-15.92 10.77-25.43 13.32-21.07 5.64-16.45 3.18-25.12 11.85-13.79 13.78-32.12 21.37-51.62 21.37-12.44 0-24.47-3.55-35.31-9.58L252 502.04c4.39 10.77 18.44 13.4 26.43 4.96l36.25-38.28 52.69 2.01c11.62.44 19.82-11.27 15.43-22.03zM263 340c15.28-15.55 17.03-14.21 38.79-20.14 13.89-3.79 24.75-14.84 28.47-28.98 7.48-28.4 5.54-24.97 25.95-45.75 10.17-10.35 14.14-25.44 10.42-39.58-7.47-28.38-7.48-24.42 0-52.83 3.72-14.14-.25-29.23-10.42-39.58-20.41-20.78-18.47-17.36-25.95-45.75-3.72-14.14-14.58-25.19-28.47-28.98-27.88-7.61-24.52-5.62-44.95-26.41-10.17-10.35-25-14.4-38.89-10.61-27.87 7.6-23.98 7.61-51.9 0-13.89-3.79-28.72.25-38.89 10.61-20.41 20.78-17.05 18.8-44.94 26.41-13.89 3.79-24.75 14.84-28.47 28.98-7.47 28.39-5.54 24.97-25.95 45.75-10.17 10.35-14.15 25.44-10.42 39.58 7.47 28.36 7.48 24.4 0 52.82-3.72 14.14.25 29.23 10.42 39.59 20.41 20.78 18.47 17.35 25.95 45.75 3.72 14.14 14.58 25.19 28.47 28.98C104.6 325.96 106.27 325 121 340c13.23 13.47 33.84 15.88 49.74 5.82a39.676 39.676 0 0 1 42.53 0c15.89 10.06 36.5 7.65 49.73-5.82zM97.66 175.96c0-53.03 42.24-96.02 94.34-96.02s94.34 42.99 94.34 96.02-42.24 96.02-94.34 96.02-94.34-42.99-94.34-96.02z"
          />
        </svg>
         
        1000
      </div>
    </div>
  </div>
</div>
`;
